compiler/src/iree/compiler/Codegen/LLVMGPU/LLVMGPUVectorLowering.cpp

```diff
@@ -82,6 +82,66 @@ struct PromoteContractOperands final
   }
 };
 
+/// true: vector
+/// false: vector
+/// pred: i1
+///
+/// select(pred, true, false) -> broadcast(pred)
+/// select(pred, false, true) -> broadcast(not(pred))
+///
+/// Ideally, this would be a canonicalization pattern on arith::SelectOp, but
+/// we cannot have arith depending on vector. Also, it would implicitly force
+/// users only using arith and vector dialect to use vector dialect. Since
+/// upstream is not willing to have this pattern anywhere, we manually add it
+/// where it is needed.
```

Comment:
You can add dialect canonicalizations that conceptually don't belong to an op but are somehow tied to a specific dialect. Or alternatively this could just be a canonicalization on vector.broadcast. Adding here is fine, but the comment reads as a jab at upstream :/

Reply:
If this is going to live in IREE, is there a good place for it in the `Codegen/Common` directory? It seems like quite a generic simplification.

Reply:
I have more general plans for canonicalizations like this by having a IREE specific interface, but that is a bit longer term work. Right now, this pattern is needed to lower maskedloads properly.